src/core/project/qgsproject.cpp

```diff
@@ -3272,20 +3281,47 @@ QStringList QgsProject::nonIdentifiableLayers() const
 
 bool QgsProject::autoTransaction() const
 {
-  return mAutoTransaction;
+  return mTransactionMode == Qgis::TransactionMode::AutomaticGroups;
 }
 
 void QgsProject::setAutoTransaction( bool autoTransaction )
 {
-  if ( autoTransaction != mAutoTransaction )
-  {
-    mAutoTransaction = autoTransaction;
+  if ( autoTransaction
+       && mTransactionMode == Qgis::TransactionMode::AutomaticGroups )
+    return;
 
-    if ( autoTransaction )
-      onMapLayersAdded( mapLayers().values() );
-    else
-      cleanTransactionGroups( true );
+  if ( ! autoTransaction
+       && mTransactionMode == Qgis::TransactionMode::None )
+    return;
+
+  if ( autoTransaction )
+  {
+    mTransactionMode = Qgis::TransactionMode::AutomaticGroups;
+    onMapLayersAdded( mapLayers().values() );
   }
+  else
+  {
+    mTransactionMode = Qgis::TransactionMode::None;
+    cleanTransactionGroups( true );
+  }
+}
+
+Qgis::TransactionMode QgsProject::transactionMode() const
+{
+  return mTransactionMode;
+}
+
+void QgsProject::setTransactionMode( Qgis::TransactionMode transactionMode )
+{
+  if ( transactionMode == mTransactionMode )
+    return;
+
+  mTransactionMode = transactionMode;
+
+  if ( mTransactionMode == Qgis::TransactionMode::AutomaticGroups )
```

Comment:
This should be a switch statement (with no default case, so warnings are raised if new values are ever added to the enum)